// @page-agent/ui/dist/lib/page-agent-ui.js
(function() {
  "use strict";
  try {
    if (typeof document != "undefined") {
      var elementStyle = document.createElement("style");
      elementStyle.appendChild(document.createTextNode("._wrapper_gtdpc_1 {\n	position: fixed;\n	bottom: 100px;\n	left: 50%;\n	transform: translateX(-50%) translateY(20px);\n	opacity: 0;\n	z-index: 2147483642; /* 比 SimulatorMask 高一层 */\n	box-sizing: border-box;\n\n	overflow: visible;\n\n	* {\n		box-sizing: border-box;\n	}\n\n	--width: 360px;\n	--height: 40px;\n	--border-radius: 12px;\n\n	--side-space: 12px; /* 控制栏两侧的间距 */\n	--history-width: calc(var(--width) - var(--side-space) * 2);\n\n	--color-1: rgb(57, 182, 255);\n	--color-2: rgb(189, 69, 251);\n	--color-3: rgb(255, 87, 51);\n	--color-4: rgb(255, 214, 0);\n\n	width: var(--width);\n	height: var(--height);\n\n	transition: all 0.3s ease-in-out;\n\n	/* 响应式设计 */\n	@media (max-width: 480px) {\n		width: calc(100vw - 40px);\n		--width: calc(100vw - 40px);\n	}\n\n	._background_gtdpc_39 {\n		position: absolute;\n		inset: -2px -8px;\n		border-radius: calc(var(--border-radius) + 4px);\n		filter: blur(16px);\n		overflow: hidden;\n		/* mix-blend-mode: lighten; */\n		/* display: none; */\n\n		&::before {\n			content: '';\n			z-index: -1;\n			pointer-events: none;\n			position: absolute;\n			width: 100%;\n			height: 100%;\n			/* left: -100%; */\n			left: 0;\n			top: 0;\n\n			background-image: linear-gradient(\n				to bottom left,\n				var(--color-1),\n				var(--color-2),\n				var(--color-1)\n			);\n			animation: _mask-running_gtdpc_1 2s linear infinite;\n		}\n		&::after {\n			content: '';\n			z-index: -1;\n			pointer-events: none;\n			position: absolute;\n			width: 100%;\n			height: 100%;\n			left: 0;\n			top: 0;\n\n			background-image: linear-gradient(\n				to bottom left,\n				var(--color-2),\n				var(--color-1),\n				var(--color-2)\n			);\n			animation: _mask-running_gtdpc_1 2s linear infinite;\n			animation-delay: 1s;\n		}\n	}\n}\n\n@keyframes _mask-running_gtdpc_1 {\n	from {\n		transform: translateX(-100%);\n	}\n	to {\n		transform: translateX(100%);\n	}\n}\n\n/* 控制栏 */\n._header_gtdpc_99 {\n	display: flex;\n	align-items: center;\n	justify-content: space-between;\n	padding: 8px 12px;\n	user-select: none;\n\n	position: absolute;\n	inset: 0;\n\n	cursor: pointer;\n	flex-shrink: 0; /* 防止 header 被压缩 */\n\n	background: rgba(0, 0, 0, 0.5);\n	backdrop-filter: blur(10px);\n	border-radius: var(--border-radius);\n	background-clip: padding-box;\n\n	box-shadow:\n		0 0 0px 2px rgba(255, 255, 255, 0.4),\n		0 0 5px 1px rgba(255, 255, 255, 0.3);\n\n	._statusSection_gtdpc_121 {\n		display: flex;\n		align-items: center;\n		gap: 8px;\n		flex: 1;\n		min-height: 24px; /* 确保垂直居中 */\n\n		._indicator_gtdpc_128 {\n			width: 6px;\n			height: 6px;\n			border-radius: 50%;\n			background: rgba(255, 255, 255, 0.5);\n			flex-shrink: 0;\n			animation: none; /* 默认无动画 */\n\n			/* 运行状态 - 有动画 */\n			&._thinking_gtdpc_137 {\n				background: rgb(57, 182, 255);\n				animation: _pulse_gtdpc_1 0.8s ease-in-out infinite;\n			}\n\n			&._tool_executing_gtdpc_142 {\n				background: rgb(189, 69, 251);\n				animation: _pulse_gtdpc_1 0.6s ease-in-out infinite;\n			}\n\n			&._retry_gtdpc_147 {\n				background: rgb(255, 214, 0);\n				animation: _retryPulse_gtdpc_1 1s ease-in-out infinite;\n			}\n\n			/* 静止状态 - 无动画 */\n			&._completed_gtdpc_153,\n			&._input_gtdpc_154,\n			&._output_gtdpc_155 {\n				background: rgb(34, 197, 94);\n				animation: none;\n			}\n\n			&._error_gtdpc_160 {\n				background: rgb(239, 68, 68);\n				animation: none;\n			}\n		}\n\n		._statusText_gtdpc_166 {\n			color: white;\n			font-size: 12px;\n			line-height: 1;\n			font-weight: 500;\n			transition: all 0.3s ease-in-out;\n			position: relative;\n			overflow: hidden;\n			display: flex;\n			align-items: center;\n			min-height: 24px; /* 确保垂直居中 */\n\n			&._fadeOut_gtdpc_178 {\n				animation: _statusTextFadeOut_gtdpc_1 0.3s ease forwards;\n			}\n\n			&._fadeIn_gtdpc_182 {\n				animation: _statusTextFadeIn_gtdpc_1 0.3s ease forwards;\n			}\n		}\n	}\n\n	._controls_gtdpc_188 {\n		display: flex;\n		align-items: center;\n		gap: 4px;\n\n		._controlButton_gtdpc_193 {\n			width: 24px;\n			height: 24px;\n			border: none;\n			border-radius: 4px;\n			background: rgba(255, 255, 255, 0.1);\n			color: white;\n			cursor: pointer;\n			display: flex;\n			align-items: center;\n			justify-content: center;\n			font-size: 12px;\n			line-height: 1;\n\n			&:hover {\n				background: rgba(255, 255, 255, 0.2);\n			}\n		}\n\n		._stopButton_gtdpc_212 {\n			background: rgba(239, 68, 68, 0.2);\n			color: rgb(255, 41, 41);\n			font-weight: 600;\n\n			&:hover {\n				background: rgba(239, 68, 68, 0.3);\n			}\n		}\n	}\n}\n\n@keyframes _statusTextFadeIn_gtdpc_1 {\n	0% {\n		opacity: 0;\n		transform: translateY(5px);\n	}\n	100% {\n		opacity: 1;\n		transform: translateY(0);\n	}\n}\n\n@keyframes _statusTextFadeOut_gtdpc_1 {\n	0% {\n		opacity: 1;\n		transform: translateY(0);\n	}\n	100% {\n		opacity: 0;\n		transform: translateY(-5px);\n	}\n}\n\n._historySectionWrapper_gtdpc_246 {\n	position: absolute;\n	width: var(--history-width);\n	bottom: var(--height);\n	left: var(--side-space);\n	z-index: -2;\n\n	padding-top: 0px;\n	visibility: collapse;\n	overflow: hidden;\n\n	transition: all 0.2s;\n\n	background: rgba(2, 0, 20, 0.5);\n	/* background: rgba(186, 186, 186, 0.2); */\n	backdrop-filter: blur(10px);\n\n	text-shadow: 0 0 1px rgba(0, 0, 0, 0.2);\n\n	border-top-left-radius: calc(var(--border-radius) + 4px);\n	border-top-right-radius: calc(var(--border-radius) + 4px);\n\n	/* border: 2px solid rgba(255, 255, 255, 0.8); */\n	border: 2px solid rgba(255, 255, 255, 0.4);\n	box-shadow: 0 4px 16px rgba(0, 0, 0, 0.6);\n\n	/* @media (prefers-color-scheme: dark) {\n		box-shadow:\n			0 8px 32px 0 rgba(0, 0, 0, 0.85),\n			0 2px 12px 0 rgba(57, 182, 255, 0.1);\n	} */\n\n	._expanded_gtdpc_278 & {\n		padding-top: 8px;\n		visibility: visible;\n	}\n\n	._historySection_gtdpc_246 {\n		position: relative;\n		overflow-y: auto;\n		overscroll-behavior: contain;\n		scrollbar-width: none;\n		max-height: 0;\n		padding-inline: 8px;\n\n		transition: max-height 0.2s;\n\n		._expanded_gtdpc_278 & {\n			max-height: 400px;\n		}\n\n		._historyItem_gtdpc_297 {\n			/* backdrop-filter: blur(10px); */\n			padding: 8px 10px;\n			margin-bottom: 6px;\n			background: linear-gradient(135deg, rgba(255, 255, 255, 0.08), rgba(255, 255, 255, 0.03));\n			border-radius: 8px;\n			border-left: 2px solid rgba(57, 182, 255, 0.5);\n			font-size: 12px;\n			color: white;\n			/* color: black; */\n			line-height: 1.3;\n			position: relative;\n			overflow: hidden;\n\n			/* 微妙的内阴影 */\n			box-shadow:\n				inset 0 1px 0 rgba(255, 255, 255, 0.1),\n				0 1px 3px rgba(0, 0, 0, 0.1);\n\n			&::before {\n				content: '';\n				position: absolute;\n				top: 0;\n				left: 0;\n				right: 0;\n				height: 1px;\n				background: linear-gradient(90deg, transparent, rgba(255, 255, 255, 0.2), transparent);\n			}\n\n			&:hover {\n				background: linear-gradient(135deg, rgba(255, 255, 255, 0.12), rgba(255, 255, 255, 0.06));\n				/* transform: translateY(-1px); */\n				box-shadow:\n					inset 0 1px 0 rgba(255, 255, 255, 0.15),\n					0 2px 4px rgba(0, 0, 0, 0.15);\n			}\n\n			&:last-child {\n				margin-bottom: 10px;\n			}\n\n			&._completed_gtdpc_153,\n			&._input_gtdpc_154,\n			&._output_gtdpc_155 {\n				border-left-color: rgb(34, 197, 94);\n				background: linear-gradient(135deg, rgba(34, 197, 94, 0.1), rgba(34, 197, 94, 0.05));\n			}\n\n			&._error_gtdpc_160 {\n				border-left-color: rgb(239, 68, 68);\n				background: linear-gradient(135deg, rgba(239, 68, 68, 0.1), rgba(239, 68, 68, 0.05));\n			}\n\n			&._retry_gtdpc_147 {\n				border-left-color: rgb(255, 214, 0);\n				background: linear-gradient(135deg, rgba(255, 214, 0, 0.1), rgba(255, 214, 0, 0.05));\n			}\n\n			&._observation_gtdpc_355 {\n				border-left-color: rgb(147, 51, 234);\n				background: linear-gradient(135deg, rgba(147, 51, 234, 0.1), rgba(147, 51, 234, 0.05));\n			}\n\n			&._question_gtdpc_360 {\n				border-left-color: rgb(255, 159, 67);\n				background: linear-gradient(135deg, rgba(255, 159, 67, 0.15), rgba(255, 159, 67, 0.08));\n			}\n\n			/* 突出显示 done 成功结果 */\n			&._doneSuccess_gtdpc_366 {\n				background: linear-gradient(\n					135deg,\n					rgba(34, 197, 94, 0.25),\n					rgba(34, 197, 94, 0.15),\n					rgba(34, 197, 94, 0.08)\n				);\n				border: none;\n				border-left: 4px solid rgb(34, 197, 94);\n				box-shadow:\n					0 4px 12px rgba(34, 197, 94, 0.3),\n					inset 0 1px 0 rgba(255, 255, 255, 0.2),\n					0 0 20px rgba(34, 197, 94, 0.1);\n				font-weight: 600;\n				color: rgb(220, 252, 231);\n				padding: 10px 12px;\n				margin-bottom: 8px;\n				border-radius: 8px;\n				position: relative;\n				overflow: hidden;\n\n				&::before {\n					background: linear-gradient(90deg, transparent, rgba(34, 197, 94, 0.4), transparent);\n				}\n\n				&::after {\n					content: '';\n					position: absolute;\n					top: 0;\n					left: -100%;\n					width: 100%;\n					height: 100%;\n					background: linear-gradient(90deg, transparent, rgba(255, 255, 255, 0.1), transparent);\n					animation: _shimmer_gtdpc_1 2s ease-in-out infinite;\n				}\n\n				._historyContent_gtdpc_402 {\n					._statusIcon_gtdpc_403 {\n						font-size: 16px;\n						animation: _celebrate_gtdpc_1 0.8s ease-in-out;\n						filter: drop-shadow(0 2px 4px rgba(34, 197, 94, 0.5));\n					}\n				}\n			}\n\n			/* 突出显示 done 失败结果 */\n			&._doneError_gtdpc_412 {\n				background: linear-gradient(\n					135deg,\n					rgba(239, 68, 68, 0.25),\n					rgba(239, 68, 68, 0.15),\n					rgba(239, 68, 68, 0.08)\n				);\n				border: none;\n				border-left: 4px solid rgb(239, 68, 68);\n				box-shadow:\n					0 4px 12px rgba(239, 68, 68, 0.3),\n					inset 0 1px 0 rgba(255, 255, 255, 0.2),\n					0 0 20px rgba(239, 68, 68, 0.1);\n				font-weight: 600;\n				color: rgb(254, 226, 226);\n				padding: 10px 12px;\n				margin-bottom: 8px;\n				border-radius: 8px;\n				position: relative;\n				overflow: hidden;\n\n				&::before {\n					background: linear-gradient(90deg, transparent, rgba(239, 68, 68, 0.4), transparent);\n				}\n\n				._historyContent_gtdpc_402 {\n					._statusIcon_gtdpc_403 {\n						font-size: 16px;\n						filter: drop-shadow(0 2px 4px rgba(239, 68, 68, 0.5));\n					}\n				}\n			}\n\n			._historyContent_gtdpc_402 {\n				display: flex;\n				align-items: flex-start;\n				gap: 8px;\n\n				word-break: break-all;\n				white-space: pre-wrap;\n\n				/* overflow-x: auto; */\n\n				._statusIcon_gtdpc_403 {\n					font-size: 12px;\n					flex-shrink: 0;\n					line-height: 1;\n					transition: all 0.3s ease;\n				}\n\n				._reflectionLines_gtdpc_462 {\n					display: flex;\n					flex-direction: column;\n					gap: 4px;\n				}\n			}\n\n			._historyMeta_gtdpc_469 {\n				font-size: 10px;\n				color: rgba(255, 255, 255, 0.6);\n				/* color: rgb(61, 61, 61); */\n				margin-top: 8px;\n				line-height: 1;\n			}\n		}\n	}\n}\n\n/* 动画关键帧 - 更快的闪烁 */\n@keyframes _pulse_gtdpc_1 {\n	0%,\n	100% {\n		opacity: 1;\n		transform: scale(1);\n	}\n	50% {\n		opacity: 0.4;\n		transform: scale(1.3);\n	}\n}\n\n/* 重试动画 - 旋转脉冲 */\n@keyframes _retryPulse_gtdpc_1 {\n	0%,\n	100% {\n		opacity: 1;\n		transform: scale(1) rotate(0deg);\n	}\n	25% {\n		opacity: 0.6;\n		transform: scale(1.2) rotate(90deg);\n	}\n	50% {\n		opacity: 0.8;\n		transform: scale(1.1) rotate(180deg);\n	}\n	75% {\n		opacity: 0.6;\n		transform: scale(1.2) rotate(270deg);\n	}\n}\n\n/* 庆祝动画 */\n@keyframes _celebrate_gtdpc_1 {\n	0%,\n	100% {\n		transform: scale(1);\n	}\n	25% {\n		transform: scale(1.2) rotate(-5deg);\n	}\n	75% {\n		transform: scale(1.2) rotate(5deg);\n	}\n}\n\n/* done 卡片的光泽效果 */\n@keyframes _shimmer_gtdpc_1 {\n	0% {\n		left: -100%;\n	}\n	100% {\n		left: 100%;\n	}\n}\n\n/* 输入区域样式 */\n._inputSectionWrapper_gtdpc_539 {\n	position: absolute;\n	width: var(--history-width);\n	top: var(--height);\n	left: var(--side-space);\n	z-index: -1;\n\n	visibility: visible;\n	overflow: hidden;\n\n	height: 48px;\n\n	transition: all 0.2s;\n\n	background: rgba(186, 186, 186, 0.2);\n	backdrop-filter: blur(10px);\n\n	border-bottom-left-radius: calc(var(--border-radius) + 4px);\n	border-bottom-right-radius: calc(var(--border-radius) + 4px);\n\n	border: 2px solid rgba(255, 255, 255, 0.3);\n	box-shadow: 0 1px 16px rgba(0, 0, 0, 0.4);\n\n	&._hidden_gtdpc_562 {\n		visibility: collapse;\n		height: 0;\n	}\n\n	._inputSection_gtdpc_539 {\n		display: flex;\n		align-items: center;\n		gap: 4px;\n		padding: 8px 8px;\n\n		._taskInput_gtdpc_573 {\n			flex: 1;\n			background: rgba(255, 255, 255, 0.4);\n			border: 1px solid rgba(255, 255, 255, 0.3);\n			border-radius: 10px;\n			padding-inline: 10px;\n			color: rgb(20, 20, 20);\n			font-size: 12px;\n			height: 28px;\n			line-height: 1;\n			outline: none;\n			transition: all 0.2s ease;\n\n			/* text-shadow: 0 0 2px rgba(255, 255, 255, 0.8); */\n\n			/* border-color: rgba(57, 182, 255, 0.3); */\n\n			&::placeholder {\n				color: rgb(53, 53, 53);\n			}\n\n			&:focus {\n				background: rgba(255, 255, 255, 0.8);\n				border-color: rgba(57, 182, 255, 0.6);\n				box-shadow: 0 0 0 2px rgba(57, 182, 255, 0.2);\n			}\n		}\n	}\n}"));
      document.head.appendChild(elementStyle);
    }
  } catch (e) {
    console.error("vite-plugin-css-injected-by-js", e);
  }
})();
var __defProp = Object.defineProperty;
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var __privateMethod = (obj, member, method) => (__accessCheck(obj, member, "access private method"), method);
var _wrapper, _indicator, _statusText, _historySection, _expandButton, _actionButton, _inputSection, _taskInput, _agent, _config, _isExpanded, _i18n, _userAnswerResolver, _isWaitingForUserAnswer, _headerUpdateTimer, _pendingHeaderText, _isAnimating, _onStatusChange, _onHistoryChange, _onActivity, _onAgentDispose, _Panel_instances, handleStatusChange_fn, handleHistoryChange_fn, handleActivity_fn, askUser_fn, getToolExecutingText_fn, handleActionButton_fn, submitTask_fn, handleUserAnswer_fn, showInputArea_fn, hideInputArea_fn, shouldShowInputArea_fn, createWrapper_fn, setupEventListeners_fn, toggle_fn, expand_fn, collapse_fn, startHeaderUpdateLoop_fn, stopHeaderUpdateLoop_fn, checkAndUpdateHeader_fn, animateTextChange_fn, updateStatusIndicator_fn, scrollToBottom_fn, renderHistory_fn, createTaskCard_fn, createHistoryCards_fn, createActionCards_fn;
const enUS = {
  ui: {
    panel: {
      ready: "Ready",
      thinking: "Thinking...",
      taskInput: "Enter new task, describe steps in detail, press Enter to submit",
      userAnswerPrompt: "Please answer the question above, press Enter to submit",
      taskTerminated: "Task terminated",
      taskCompleted: "Task completed",
      userAnswer: "User answer: {{input}}",
      question: "Question: {{question}}",
      waitingPlaceholder: "Waiting for task to start...",
      stop: "Stop",
      close: "Close",
      expand: "Expand history",
      collapse: "Collapse history",
      step: "Step {{number}}"
    },
    tools: {
      clicking: "Clicking element [{{index}}]...",
      inputting: "Inputting text to element [{{index}}]...",
      selecting: 'Selecting option "{{text}}"...',
      scrolling: "Scrolling page...",
      waiting: "Waiting {{seconds}} seconds...",
      askingUser: "Asking user...",
      done: "Task done",
      clicked: "🖱️ Clicked element [{{index}}]",
      inputted: '⌨️ Inputted text "{{text}}"',
      selected: '☑️ Selected option "{{text}}"',
      scrolled: "🛞 Page scrolled",
      waited: "⌛️ Wait completed",
      executing: "Executing {{toolName}}...",
      resultSuccess: "success",
      resultFailure: "failed",
      resultError: "error"
    },
    errors: {
      elementNotFound: "No interactive element found at index {{index}}",
      taskRequired: "Task description is required",
      executionFailed: "Task execution failed",
      notInputElement: "Element is not an input or textarea",
      notSelectElement: "Element is not a select element",
      optionNotFound: 'Option "{{text}}" not found'
    }
  }
};
const zhCN = {
  ui: {
    panel: {
      ready: "准备就绪",
      thinking: "正在思考...",
      taskInput: "输入新任务，详细描述步骤，回车提交",
      userAnswerPrompt: "请回答上面问题，回车提交",
      taskTerminated: "任务已终止",
      taskCompleted: "任务结束",
      userAnswer: "用户回答: {{input}}",
      question: "询问: {{question}}",
      waitingPlaceholder: "等待任务开始...",
      stop: "终止",
      close: "关闭",
      expand: "展开历史",
      collapse: "收起历史",
      step: "步骤 {{number}}"
    },
    tools: {
      clicking: "正在点击元素 [{{index}}]...",
      inputting: "正在输入文本到元素 [{{index}}]...",
      selecting: '正在选择选项 "{{text}}"...',
      scrolling: "正在滚动页面...",
      waiting: "等待 {{seconds}} 秒...",
      askingUser: "正在询问用户...",
      done: "结束任务",
      clicked: "🖱️ 已点击元素 [{{index}}]",
      inputted: '⌨️ 已输入文本 "{{text}}"',
      selected: '☑️ 已选择选项 "{{text}}"',
      scrolled: "🛞 页面滚动完成",
      waited: "⌛️ 等待完成",
      executing: "正在执行 {{toolName}}...",
      resultSuccess: "成功",
      resultFailure: "失败",
      resultError: "错误"
    },
    errors: {
      elementNotFound: "未找到索引为 {{index}} 的交互元素",
      taskRequired: "任务描述不能为空",
      executionFailed: "任务执行失败",
      notInputElement: "元素不是输入框或文本域",
      notSelectElement: "元素不是选择框",
      optionNotFound: '未找到选项 "{{text}}"'
    }
  }
};
const locales = {
  "en-US": enUS,
  "zh-CN": zhCN
};
const _I18n = class _I18n {
  language;
  translations;
  constructor(language = "en-US") {
    this.language = language in locales ? language : "en-US";
    this.translations = locales[this.language];
  }
  // 类型安全的翻译方法
  t(key, params) {
    const value = this.getNestedValue(this.translations, key);
    if (!value) {
      console.warn(`Translation key "${key}" not found for language "${this.language}"`);
      return key;
    }
    if (params) {
      return this.interpolate(value, params);
    }
    return value;
  }
  getNestedValue(obj, path) {
    return path.split(".").reduce((current, key) => current?.[key], obj);
  }
  interpolate(template, params) {
    return template.replace(/\{\{(\w+)\}\}/g, (match, key) => {
      return params[key] != null ? params[key].toString() : match;
    });
  }
  getLanguage() {
    return this.language;
  }
};
__name(_I18n, "I18n");
let I18n = _I18n;
function truncate(text, maxLength) {
  if (text.length > maxLength) {
    return text.substring(0, maxLength) + "...";
  }
  return text;
}
__name(truncate, "truncate");
function escapeHtml(text) {
  return text.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#039;");
}
__name(escapeHtml, "escapeHtml");
const wrapper = "_wrapper_gtdpc_1";
const background = "_background_gtdpc_39";
const header = "_header_gtdpc_99";
const pulse = "_pulse_gtdpc_1";
const retryPulse = "_retryPulse_gtdpc_1";
const statusTextFadeOut = "_statusTextFadeOut_gtdpc_1";
const statusTextFadeIn = "_statusTextFadeIn_gtdpc_1";
const statusSection = "_statusSection_gtdpc_121";
const indicator = "_indicator_gtdpc_128";
const thinking = "_thinking_gtdpc_137";
const tool_executing = "_tool_executing_gtdpc_142";
const retry = "_retry_gtdpc_147";
const completed = "_completed_gtdpc_153";
const input = "_input_gtdpc_154";
const output = "_output_gtdpc_155";
const error = "_error_gtdpc_160";
const statusText = "_statusText_gtdpc_166";
const fadeOut = "_fadeOut_gtdpc_178";
const fadeIn = "_fadeIn_gtdpc_182";
const controls = "_controls_gtdpc_188";
const controlButton = "_controlButton_gtdpc_193";
const stopButton = "_stopButton_gtdpc_212";
const historySectionWrapper = "_historySectionWrapper_gtdpc_246";
const shimmer = "_shimmer_gtdpc_1";
const celebrate = "_celebrate_gtdpc_1";
const expanded = "_expanded_gtdpc_278";
const historySection = "_historySection_gtdpc_246";
const historyItem = "_historyItem_gtdpc_297";
const observation = "_observation_gtdpc_355";
const question = "_question_gtdpc_360";
const doneSuccess = "_doneSuccess_gtdpc_366";
const historyContent = "_historyContent_gtdpc_402";
const statusIcon = "_statusIcon_gtdpc_403";
const doneError = "_doneError_gtdpc_412";
const reflectionLines = "_reflectionLines_gtdpc_462";
const historyMeta = "_historyMeta_gtdpc_469";
const inputSectionWrapper = "_inputSectionWrapper_gtdpc_539";
const hidden = "_hidden_gtdpc_562";
const inputSection = "_inputSection_gtdpc_539";
const taskInput = "_taskInput_gtdpc_573";
const styles = {
  wrapper,
  "mask-running": "_mask-running_gtdpc_1",
  background,
  header,
  pulse,
  retryPulse,
  statusTextFadeOut,
  statusTextFadeIn,
  statusSection,
  indicator,
  thinking,
  tool_executing,
  retry,
  completed,
  input,
  output,
  error,
  statusText,
  fadeOut,
  fadeIn,
  controls,
  controlButton,
  stopButton,
  historySectionWrapper,
  shimmer,
  celebrate,
  expanded,
  historySection,
  historyItem,
  observation,
  question,
  doneSuccess,
  historyContent,
  statusIcon,
  doneError,
  reflectionLines,
  historyMeta,
  inputSectionWrapper,
  hidden,
  inputSection,
  taskInput
};
function createCard({ icon, content, meta, type }) {
  const typeClass = type ? styles[type] : "";
  const contentHtml = Array.isArray(content) ? `<div class="${styles.reflectionLines}">${content.map((line) => `<span>${escapeHtml(line)}</span>`).join("")}</div>` : `<span>${escapeHtml(content)}</span>`;
  return `
		<div class="${styles.historyItem} ${typeClass}">
			<div class="${styles.historyContent}">
				<span class="${styles.statusIcon}">${icon}</span>
				${contentHtml}
			</div>
			${meta ? `<div class="${styles.historyMeta}">${meta}</div>` : ""}
		</div>
	`;
}
__name(createCard, "createCard");
function createReflectionLines(reflection) {
  const lines = [];
  if (reflection.evaluation_previous_goal) {
    lines.push(`🔍 ${reflection.evaluation_previous_goal}`);
  }
  if (reflection.memory) {
    lines.push(`💾 ${reflection.memory}`);
  }
  if (reflection.next_goal) {
    lines.push(`🎯 ${reflection.next_goal}`);
  }
  return lines;
}
__name(createReflectionLines, "createReflectionLines");
const _Panel = class _Panel {
  /**
   * Create a Panel bound to an agent
   * @param agent - Agent instance that implements PanelAgentAdapter
   * @param config - Optional panel configuration
   */
  constructor(agent, config = {}) {
    __privateAdd(this, _Panel_instances);
    __privateAdd(this, _wrapper);
    __privateAdd(this, _indicator);
    __privateAdd(this, _statusText);
    __privateAdd(this, _historySection);
    __privateAdd(this, _expandButton);
    __privateAdd(this, _actionButton);
    __privateAdd(this, _inputSection);
    __privateAdd(this, _taskInput);
    __privateAdd(this, _agent);
    __privateAdd(this, _config);
    __privateAdd(this, _isExpanded, false);
    __privateAdd(this, _i18n);
    __privateAdd(this, _userAnswerResolver, null);
    __privateAdd(this, _isWaitingForUserAnswer, false);
    __privateAdd(this, _headerUpdateTimer, null);
    __privateAdd(this, _pendingHeaderText, null);
    __privateAdd(this, _isAnimating, false);
    // Event handlers (bound for removal)
    __privateAdd(this, _onStatusChange, /* @__PURE__ */ __name(() => __privateMethod(this, _Panel_instances, handleStatusChange_fn).call(this), "#onStatusChange"));
    __privateAdd(this, _onHistoryChange, /* @__PURE__ */ __name(() => __privateMethod(this, _Panel_instances, handleHistoryChange_fn).call(this), "#onHistoryChange"));
    __privateAdd(this, _onActivity, /* @__PURE__ */ __name((e) => __privateMethod(this, _Panel_instances, handleActivity_fn).call(this, e.detail), "#onActivity"));
    __privateAdd(this, _onAgentDispose, /* @__PURE__ */ __name(() => this.dispose(), "#onAgentDispose"));
    __privateSet(this, _agent, agent);
    __privateSet(this, _config, config);
    __privateSet(this, _i18n, new I18n(config.language ?? "en-US"));
    __privateGet(this, _agent).onAskUser = (question2) => __privateMethod(this, _Panel_instances, askUser_fn).call(this, question2);
    __privateSet(this, _wrapper, __privateMethod(this, _Panel_instances, createWrapper_fn).call(this));
    __privateSet(this, _indicator, __privateGet(this, _wrapper).querySelector(`.${styles.indicator}`));
    __privateSet(this, _statusText, __privateGet(this, _wrapper).querySelector(`.${styles.statusText}`));
    __privateSet(this, _historySection, __privateGet(this, _wrapper).querySelector(`.${styles.historySection}`));
    __privateSet(this, _expandButton, __privateGet(this, _wrapper).querySelector(`.${styles.expandButton}`));
    __privateSet(this, _actionButton, __privateGet(this, _wrapper).querySelector(`.${styles.stopButton}`));
    __privateSet(this, _inputSection, __privateGet(this, _wrapper).querySelector(`.${styles.inputSectionWrapper}`));
    __privateSet(this, _taskInput, __privateGet(this, _wrapper).querySelector(`.${styles.taskInput}`));
    __privateGet(this, _agent).addEventListener("statuschange", __privateGet(this, _onStatusChange));
    __privateGet(this, _agent).addEventListener("historychange", __privateGet(this, _onHistoryChange));
    __privateGet(this, _agent).addEventListener("activity", __privateGet(this, _onActivity));
    __privateGet(this, _agent).addEventListener("dispose", __privateGet(this, _onAgentDispose));
    __privateMethod(this, _Panel_instances, setupEventListeners_fn).call(this);
    __privateMethod(this, _Panel_instances, startHeaderUpdateLoop_fn).call(this);
    __privateMethod(this, _Panel_instances, showInputArea_fn).call(this);
    this.hide();
  }
  get wrapper() {
    return __privateGet(this, _wrapper);
  }
  // ========== Public control methods ==========
  show() {
    this.wrapper.style.display = "block";
    void this.wrapper.offsetHeight;
    this.wrapper.style.opacity = "1";
    this.wrapper.style.transform = "translateX(-50%) translateY(0)";
  }
  hide() {
    this.wrapper.style.opacity = "0";
    this.wrapper.style.transform = "translateX(-50%) translateY(20px)";
    this.wrapper.style.display = "none";
  }
  reset() {
    __privateGet(this, _statusText).textContent = __privateGet(this, _i18n).t("ui.panel.ready");
    __privateMethod(this, _Panel_instances, updateStatusIndicator_fn).call(this, "thinking");
    __privateMethod(this, _Panel_instances, renderHistory_fn).call(this);
    __privateMethod(this, _Panel_instances, collapse_fn).call(this);
    __privateSet(this, _isWaitingForUserAnswer, false);
    __privateSet(this, _userAnswerResolver, null);
    __privateMethod(this, _Panel_instances, showInputArea_fn).call(this);
  }
  expand() {
    __privateMethod(this, _Panel_instances, expand_fn).call(this);
  }
  collapse() {
    __privateMethod(this, _Panel_instances, collapse_fn).call(this);
  }
  /**
   * Dispose panel and clean up event listeners
   */
  dispose() {
    __privateGet(this, _agent).removeEventListener("statuschange", __privateGet(this, _onStatusChange));
    __privateGet(this, _agent).removeEventListener("historychange", __privateGet(this, _onHistoryChange));
    __privateGet(this, _agent).removeEventListener("activity", __privateGet(this, _onActivity));
    __privateGet(this, _agent).removeEventListener("dispose", __privateGet(this, _onAgentDispose));
    __privateSet(this, _isWaitingForUserAnswer, false);
    __privateMethod(this, _Panel_instances, stopHeaderUpdateLoop_fn).call(this);
    this.wrapper.remove();
  }
};
_wrapper = new WeakMap();
_indicator = new WeakMap();
_statusText = new WeakMap();
_historySection = new WeakMap();
_expandButton = new WeakMap();
_actionButton = new WeakMap();
_inputSection = new WeakMap();
_taskInput = new WeakMap();
_agent = new WeakMap();
_config = new WeakMap();
_isExpanded = new WeakMap();
_i18n = new WeakMap();
_userAnswerResolver = new WeakMap();
_isWaitingForUserAnswer = new WeakMap();
_headerUpdateTimer = new WeakMap();
_pendingHeaderText = new WeakMap();
_isAnimating = new WeakMap();
_onStatusChange = new WeakMap();
_onHistoryChange = new WeakMap();
_onActivity = new WeakMap();
_onAgentDispose = new WeakMap();
_Panel_instances = new WeakSet();
// ========== Agent event handlers ==========
/** Handle agent status change */
handleStatusChange_fn = /* @__PURE__ */ __name(function() {
  const status = __privateGet(this, _agent).status;
  const indicatorType = status === "running" ? "thinking" : status === "idle" ? "thinking" : status;
  __privateMethod(this, _Panel_instances, updateStatusIndicator_fn).call(this, indicatorType);
  if (status === "running") {
    __privateGet(this, _actionButton).textContent = "■";
    __privateGet(this, _actionButton).title = __privateGet(this, _i18n).t("ui.panel.stop");
  } else {
    __privateGet(this, _actionButton).textContent = "X";
    __privateGet(this, _actionButton).title = __privateGet(this, _i18n).t("ui.panel.close");
  }
  if (status === "running") {
    this.show();
    __privateMethod(this, _Panel_instances, hideInputArea_fn).call(this);
  }
  if (status === "completed" || status === "error") {
    if (!__privateGet(this, _isExpanded)) {
      __privateMethod(this, _Panel_instances, expand_fn).call(this);
    }
    if (__privateMethod(this, _Panel_instances, shouldShowInputArea_fn).call(this)) {
      __privateMethod(this, _Panel_instances, showInputArea_fn).call(this);
    }
  }
}, "#handleStatusChange");
/** Handle agent history change - re-render history list from agent.history */
handleHistoryChange_fn = /* @__PURE__ */ __name(function() {
  __privateMethod(this, _Panel_instances, renderHistory_fn).call(this);
}, "#handleHistoryChange");
/**
 * Handle agent activity - transient state for immediate UI feedback
 * Activity events are NOT persisted in history, only used for header bar updates
 */
handleActivity_fn = /* @__PURE__ */ __name(function(activity) {
  switch (activity.type) {
    case "thinking":
      __privateSet(this, _pendingHeaderText, __privateGet(this, _i18n).t("ui.panel.thinking"));
      __privateMethod(this, _Panel_instances, updateStatusIndicator_fn).call(this, "thinking");
      break;
    case "executing":
      __privateSet(this, _pendingHeaderText, __privateMethod(this, _Panel_instances, getToolExecutingText_fn).call(this, activity.tool, activity.input));
      __privateMethod(this, _Panel_instances, updateStatusIndicator_fn).call(this, "executing");
      break;
    case "executed":
      __privateSet(this, _pendingHeaderText, truncate(activity.output, 50));
      break;
    case "retrying":
      __privateSet(this, _pendingHeaderText, `Retrying (${activity.attempt}/${activity.maxAttempts})`);
      __privateMethod(this, _Panel_instances, updateStatusIndicator_fn).call(this, "retrying");
      break;
    case "error":
      __privateSet(this, _pendingHeaderText, truncate(activity.message, 50));
      __privateMethod(this, _Panel_instances, updateStatusIndicator_fn).call(this, "error");
      break;
  }
}, "#handleActivity");
/**
 * Ask for user input (internal, called by agent via onAskUser)
 */
askUser_fn = /* @__PURE__ */ __name(function(question2) {
  return new Promise((resolve) => {
    __privateSet(this, _isWaitingForUserAnswer, true);
    __privateSet(this, _userAnswerResolver, resolve);
    if (!__privateGet(this, _isExpanded)) {
      __privateMethod(this, _Panel_instances, expand_fn).call(this);
    }
    const tempCard = document.createElement("div");
    tempCard.innerHTML = createCard({
      icon: "❓",
      content: `Question: ${question2}`,
      type: "question"
    });
    const cardElement = tempCard.firstElementChild;
    cardElement.setAttribute("data-temp-card", "true");
    __privateGet(this, _historySection).appendChild(cardElement);
    __privateMethod(this, _Panel_instances, scrollToBottom_fn).call(this);
    __privateMethod(this, _Panel_instances, showInputArea_fn).call(this, __privateGet(this, _i18n).t("ui.panel.userAnswerPrompt"));
  });
}, "#askUser");
// ========== Private methods ==========
getToolExecutingText_fn = /* @__PURE__ */ __name(function(toolName, args) {
  const a = args;
  switch (toolName) {
    case "click_element_by_index":
      return __privateGet(this, _i18n).t("ui.tools.clicking", { index: a.index });
    case "input_text":
      return __privateGet(this, _i18n).t("ui.tools.inputting", { index: a.index });
    case "select_dropdown_option":
      return __privateGet(this, _i18n).t("ui.tools.selecting", { text: a.text });
    case "scroll":
      return __privateGet(this, _i18n).t("ui.tools.scrolling");
    case "wait":
      return __privateGet(this, _i18n).t("ui.tools.waiting", { seconds: a.seconds });
    case "ask_user":
      return __privateGet(this, _i18n).t("ui.tools.askingUser");
    case "done":
      return __privateGet(this, _i18n).t("ui.tools.done");
    default:
      return __privateGet(this, _i18n).t("ui.tools.executing", { toolName });
  }
}, "#getToolExecutingText");
/**
 * Action button handler: stop when running, close (dispose) when idle
 */
handleActionButton_fn = /* @__PURE__ */ __name(function() {
  if (__privateGet(this, _agent).status === "running") {
    __privateGet(this, _agent).stop();
  } else {
    __privateGet(this, _agent).dispose();
  }
}, "#handleActionButton");
/**
 * Submit task
 */
submitTask_fn = /* @__PURE__ */ __name(function() {
  const input2 = __privateGet(this, _taskInput).value.trim();
  if (!input2) return;
  __privateMethod(this, _Panel_instances, hideInputArea_fn).call(this);
  if (__privateGet(this, _isWaitingForUserAnswer)) {
    __privateMethod(this, _Panel_instances, handleUserAnswer_fn).call(this, input2);
  } else {
    __privateGet(this, _agent).execute(input2);
  }
}, "#submitTask");
/**
 * Handle user answer
 */
handleUserAnswer_fn = /* @__PURE__ */ __name(function(input2) {
  Array.from(__privateGet(this, _historySection).children).forEach((child) => {
    if (child.getAttribute("data-temp-card") === "true") {
      child.remove();
    }
  });
  __privateSet(this, _isWaitingForUserAnswer, false);
  if (__privateGet(this, _userAnswerResolver)) {
    __privateGet(this, _userAnswerResolver).call(this, input2);
    __privateSet(this, _userAnswerResolver, null);
  }
}, "#handleUserAnswer");
/**
 * Show input area
 */
showInputArea_fn = /* @__PURE__ */ __name(function(placeholder) {
  __privateGet(this, _taskInput).value = "";
  __privateGet(this, _taskInput).placeholder = placeholder || __privateGet(this, _i18n).t("ui.panel.taskInput");
  __privateGet(this, _inputSection).classList.remove(styles.hidden);
  setTimeout(() => {
    __privateGet(this, _taskInput).focus();
  }, 100);
}, "#showInputArea");
/**
 * Hide input area
 */
hideInputArea_fn = /* @__PURE__ */ __name(function() {
  __privateGet(this, _inputSection).classList.add(styles.hidden);
}, "#hideInputArea");
/**
 * Check if input area should be shown
 */
shouldShowInputArea_fn = /* @__PURE__ */ __name(function() {
  if (__privateGet(this, _isWaitingForUserAnswer)) return true;
  const history = __privateGet(this, _agent).history;
  if (history.length === 0) {
    return true;
  }
  const status = __privateGet(this, _agent).status;
  const isTaskEnded = status === "completed" || status === "error";
  if (isTaskEnded) {
    return __privateGet(this, _config).promptForNextTask ?? true;
  }
  return false;
}, "#shouldShowInputArea");
createWrapper_fn = /* @__PURE__ */ __name(function() {
  const taskInputMaxLength = 1e3;
  const wrapper2 = document.createElement("div");
  wrapper2.id = "page-agent-runtime_agent-panel";
  wrapper2.className = styles.wrapper;
  wrapper2.setAttribute("data-browser-use-ignore", "true");
  wrapper2.setAttribute("data-page-agent-ignore", "true");
  wrapper2.innerHTML = `
			<div class="${styles.background}"></div>
			<div class="${styles.historySectionWrapper}">
				<div class="${styles.historySection}">
					<div class="${styles.historyItem}">
						<div class="${styles.historyContent}">
							<span class="${styles.statusIcon}">🧠</span>
							<span>${__privateGet(this, _i18n).t("ui.panel.waitingPlaceholder")}</span>
						</div>
					</div>
				</div>
			</div>
			<div class="${styles.header}">
				<div class="${styles.statusSection}">
					<div class="${styles.indicator} ${styles.thinking}"></div>
					<div class="${styles.statusText}">${__privateGet(this, _i18n).t("ui.panel.ready")}</div>
				</div>
				<div class="${styles.controls}">
					<button class="${styles.controlButton} ${styles.expandButton}" title="${__privateGet(this, _i18n).t("ui.panel.expand")}">
						▼
					</button>
					<button class="${styles.controlButton} ${styles.stopButton}" title="${__privateGet(this, _i18n).t("ui.panel.close")}">
						X
					</button>
				</div>
			</div>
			<div class="${styles.inputSectionWrapper} ${styles.hidden}">
				<div class="${styles.inputSection}">
					<input 
						type="text" 
						class="${styles.taskInput}" 
						maxlength="${taskInputMaxLength}"
					/>
				</div>
			</div>
		`;
  document.body.appendChild(wrapper2);
  return wrapper2;
}, "#createWrapper");
setupEventListeners_fn = /* @__PURE__ */ __name(function() {
  const header2 = this.wrapper.querySelector(`.${styles.header}`);
  header2.addEventListener("click", (e) => {
    if (e.target.closest(`.${styles.controlButton}`)) {
      return;
    }
    __privateMethod(this, _Panel_instances, toggle_fn).call(this);
  });
  __privateGet(this, _expandButton).addEventListener("click", (e) => {
    e.stopPropagation();
    __privateMethod(this, _Panel_instances, toggle_fn).call(this);
  });
  __privateGet(this, _actionButton).addEventListener("click", (e) => {
    e.stopPropagation();
    __privateMethod(this, _Panel_instances, handleActionButton_fn).call(this);
  });
  __privateGet(this, _taskInput).addEventListener("keydown", (e) => {
    if (e.isComposing) return;
    if (e.key === "Enter") {
      e.preventDefault();
      __privateMethod(this, _Panel_instances, submitTask_fn).call(this);
    }
  });
  __privateGet(this, _inputSection).addEventListener("click", (e) => {
    e.stopPropagation();
  });
}, "#setupEventListeners");
toggle_fn = /* @__PURE__ */ __name(function() {
  if (__privateGet(this, _isExpanded)) {
    __privateMethod(this, _Panel_instances, collapse_fn).call(this);
  } else {
    __privateMethod(this, _Panel_instances, expand_fn).call(this);
  }
}, "#toggle");
expand_fn = /* @__PURE__ */ __name(function() {
  __privateSet(this, _isExpanded, true);
  this.wrapper.classList.add(styles.expanded);
  __privateGet(this, _expandButton).textContent = "▲";
}, "#expand");
collapse_fn = /* @__PURE__ */ __name(function() {
  __privateSet(this, _isExpanded, false);
  this.wrapper.classList.remove(styles.expanded);
  __privateGet(this, _expandButton).textContent = "▼";
}, "#collapse");
/**
 * Start periodic header update loop
 */
startHeaderUpdateLoop_fn = /* @__PURE__ */ __name(function() {
  __privateSet(this, _headerUpdateTimer, setInterval(() => {
    __privateMethod(this, _Panel_instances, checkAndUpdateHeader_fn).call(this);
  }, 450));
}, "#startHeaderUpdateLoop");
/**
 * Stop periodic header update loop
 */
stopHeaderUpdateLoop_fn = /* @__PURE__ */ __name(function() {
  if (__privateGet(this, _headerUpdateTimer)) {
    clearInterval(__privateGet(this, _headerUpdateTimer));
    __privateSet(this, _headerUpdateTimer, null);
  }
}, "#stopHeaderUpdateLoop");
/**
 * Check if header needs update and trigger animation if not currently animating
 */
checkAndUpdateHeader_fn = /* @__PURE__ */ __name(function() {
  if (!__privateGet(this, _pendingHeaderText) || __privateGet(this, _isAnimating)) {
    return;
  }
  if (__privateGet(this, _statusText).textContent === __privateGet(this, _pendingHeaderText)) {
    __privateSet(this, _pendingHeaderText, null);
    return;
  }
  const textToShow = __privateGet(this, _pendingHeaderText);
  __privateSet(this, _pendingHeaderText, null);
  __privateMethod(this, _Panel_instances, animateTextChange_fn).call(this, textToShow);
}, "#checkAndUpdateHeader");
/**
 * Animate text change with fade out/in effect
 */
animateTextChange_fn = /* @__PURE__ */ __name(function(newText) {
  __privateSet(this, _isAnimating, true);
  __privateGet(this, _statusText).classList.add(styles.fadeOut);
  setTimeout(() => {
    __privateGet(this, _statusText).textContent = newText;
    __privateGet(this, _statusText).classList.remove(styles.fadeOut);
    __privateGet(this, _statusText).classList.add(styles.fadeIn);
    setTimeout(() => {
      __privateGet(this, _statusText).classList.remove(styles.fadeIn);
      __privateSet(this, _isAnimating, false);
    }, 300);
  }, 150);
}, "#animateTextChange");
updateStatusIndicator_fn = /* @__PURE__ */ __name(function(type) {
  __privateGet(this, _indicator).className = styles.indicator;
  __privateGet(this, _indicator).classList.add(styles[type]);
}, "#updateStatusIndicator");
scrollToBottom_fn = /* @__PURE__ */ __name(function() {
  setTimeout(() => {
    __privateGet(this, _historySection).scrollTop = __privateGet(this, _historySection).scrollHeight;
  }, 0);
}, "#scrollToBottom");
/**
 * Render history directly from agent.history
 *
 * Renders:
 * 1. Task (first item, from agent.task)
 * 2. Reflection cards (evaluation, memory, next_goal)
 * 3. Tool execution with output
 * 4. Observations
 */
renderHistory_fn = /* @__PURE__ */ __name(function() {
  const items = [];
  const task = __privateGet(this, _agent).task;
  if (task) {
    items.push(__privateMethod(this, _Panel_instances, createTaskCard_fn).call(this, task));
  }
  const history = __privateGet(this, _agent).history;
  for (const event of history) {
    items.push(...__privateMethod(this, _Panel_instances, createHistoryCards_fn).call(this, event));
  }
  __privateGet(this, _historySection).innerHTML = items.join("");
  __privateMethod(this, _Panel_instances, scrollToBottom_fn).call(this);
}, "#renderHistory");
createTaskCard_fn = /* @__PURE__ */ __name(function(task) {
  return createCard({ icon: "🎯", content: task, type: "input" });
}, "#createTaskCard");
/** Create cards for a history event */
createHistoryCards_fn = /* @__PURE__ */ __name(function(event) {
  const cards = [];
  const meta = event.type === "step" && event.stepIndex !== void 0 ? __privateGet(this, _i18n).t("ui.panel.step", {
    number: (event.stepIndex + 1).toString()
  }) : void 0;
  if (event.type === "step") {
    if (event.reflection) {
      const lines = createReflectionLines(event.reflection);
      if (lines.length > 0) {
        cards.push(createCard({ icon: "🧠", content: lines, meta }));
      }
    }
    const action = event.action;
    if (action) {
      cards.push(...__privateMethod(this, _Panel_instances, createActionCards_fn).call(this, action, meta));
    }
  } else if (event.type === "observation") {
    cards.push(
      createCard({ icon: "👁️", content: event.content || "", meta, type: "observation" })
    );
  } else if (event.type === "user_takeover") {
    cards.push(createCard({ icon: "👤", content: "User takeover", meta, type: "input" }));
  } else if (event.type === "retry") {
    const retryInfo = `${event.message || "Retrying"} (${event.attempt}/${event.maxAttempts})`;
    cards.push(createCard({ icon: "🔄", content: retryInfo, meta, type: "observation" }));
  } else if (event.type === "error") {
    cards.push(
      createCard({ icon: "❌", content: event.message || "Error", meta, type: "observation" })
    );
  }
  return cards;
}, "#createHistoryCards");
/** Create cards for an action */
createActionCards_fn = /* @__PURE__ */ __name(function(action, meta) {
  const cards = [];
  if (action.name === "done") {
    const input2 = action.input;
    const text = input2.text || action.output || "";
    if (text) {
      cards.push(createCard({ icon: "🤖", content: text, meta, type: "output" }));
    }
  } else if (action.name === "ask_user") {
    const input2 = action.input;
    const answer = action.output.replace(/^User answered:\s*/i, "");
    cards.push(
      createCard({
        icon: "❓",
        content: `Question: ${input2.question || ""}`,
        meta,
        type: "question"
      })
    );
    cards.push(createCard({ icon: "💬", content: `Answer: ${answer}`, meta, type: "input" }));
  } else {
    const toolText = __privateMethod(this, _Panel_instances, getToolExecutingText_fn).call(this, action.name, action.input);
    cards.push(createCard({ icon: "🔨", content: toolText, meta }));
    if (action.output?.length > 0) {
      cards.push(createCard({ icon: "🔨", content: action.output, meta, type: "output" }));
    }
  }
  return cards;
}, "#createActionCards");
__name(_Panel, "Panel");
let Panel = _Panel;
export {
  I18n,
  Panel
};
