// @page-agent/page-controller/dist/lib/SimulatorMask-BHnQ6LmL.js
(function() {
  "use strict";
  try {
    if (typeof document != "undefined") {
      var elementStyle = document.createElement("style");
      elementStyle.appendChild(document.createTextNode(`._wrapper_1ooyb_1 {
	position: fixed;
	inset: 0;
	z-index: 2147483641; /* 确保在所有元素之上，除了 panel */
	cursor: wait;
	overflow: hidden;

	display: none;
}

._wrapper_1ooyb_1._visible_1ooyb_11 {
	display: block;
}
/* AI 光标样式 */
._cursor_1dgwb_2 {
	position: absolute;
	width: var(--cursor-size, 75px);
	height: var(--cursor-size, 75px);
	pointer-events: none;
	z-index: 10000;
}

._cursorBorder_1dgwb_10 {
	position: absolute;
	width: 100%;
	height: 100%;
	background: linear-gradient(45deg, rgb(57, 182, 255), rgb(189, 69, 251));
	mask-image: url("data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%20100%20100'%20fill='none'%3e%3cg%3e%3cpath%20d='M%2015%2042%20L%2015%2036.99%20Q%2015%2031.99%2023.7%2031.99%20L%2028.05%2031.99%20Q%2032.41%2031.99%2032.41%2021.99%20L%2032.41%2017%20Q%2032.41%2012%2041.09%2016.95%20L%2076.31%2037.05%20Q%2085%2042%2076.31%2046.95%20L%2041.09%2067.05%20Q%2032.41%2072%2032.41%2062.01%20L%2032.41%2057.01%20Q%2032.41%2052.01%2023.7%2052.01%20L%2019.35%2052.01%20Q%2015%2052.01%2015%2047.01%20Z'%20fill='none'%20stroke='%23000000'%20stroke-width='6'%20stroke-miterlimit='10'%20style='stroke:%20light-dark(rgb(0,%200,%200),%20rgb(255,%20255,%20255));'/%3e%3c/g%3e%3c/svg%3e");
	mask-size: 100% 100%;
	mask-repeat: no-repeat;

	transform-origin: center;
	transform: rotate(-135deg) scale(1.2);
	margin-left: -10px;
	margin-top: -18px;
}

._cursorFilling_1dgwb_25 {
	position: absolute;
	width: 100%;
	height: 100%;
	background: url("data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%20100%20100'%3e%3cdefs%3e%3c/defs%3e%3cg%20xmlns='http://www.w3.org/2000/svg'%20style='filter:%20drop-shadow(light-dark(rgba(0,%200,%200,%200.4),%20rgba(237,%20237,%20237,%200.4))%203px%204px%204px);'%3e%3cpath%20d='M%2015%2042%20L%2015%2036.99%20Q%2015%2031.99%2023.7%2031.99%20L%2028.05%2031.99%20Q%2032.41%2031.99%2032.41%2021.99%20L%2032.41%2017%20Q%2032.41%2012%2041.09%2016.95%20L%2076.31%2037.05%20Q%2085%2042%2076.31%2046.95%20L%2041.09%2067.05%20Q%2032.41%2072%2032.41%2062.01%20L%2032.41%2057.01%20Q%2032.41%2052.01%2023.7%2052.01%20L%2019.35%2052.01%20Q%2015%2052.01%2015%2047.01%20Z'%20fill='%23ffffff'%20stroke='none'%20style='fill:%20%23ffffff;'/%3e%3c/g%3e%3c/svg%3e");
	background-size: 100% 100%;
	background-repeat: no-repeat;

	transform-origin: center;
	transform: rotate(-135deg) scale(1.2);
	margin-left: -10px;
	margin-top: -18px;
}

._cursorRipple_1dgwb_39 {
	position: absolute;
	width: 100%;
	height: 100%;
	pointer-events: none;
	margin-left: -50%;
	margin-top: -50%;

	&::after {
		content: '';
		opacity: 0;
		position: absolute;
		inset: 0;
		border: 4px solid rgba(57, 182, 255, 1);
		border-radius: 50%;
	}
}

._cursor_1dgwb_2._clicking_1dgwb_57 ._cursorRipple_1dgwb_39::after {
	animation: _cursor-ripple_1dgwb_1 300ms ease-out forwards;
}

@keyframes _cursor-ripple_1dgwb_1 {
	0% {
		transform: scale(0);
		opacity: 1;
	}
	100% {
		transform: scale(2);
		opacity: 0;
	}
}`));
      document.head.appendChild(elementStyle);
    }
  } catch (e) {
    console.error("vite-plugin-css-injected-by-js", e);
  }
})();
var __defProp = Object.defineProperty;
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var __privateMethod = (obj, member, method) => (__accessCheck(obj, member, "access private method"), method);
var _cursor, _currentCursorX, _currentCursorY, _targetCursorX, _targetCursorY, _SimulatorMask_instances, createCursor_fn, moveCursorToTarget_fn;
import { Motion } from "ai-motion";
function hasDarkModeClass() {
  const DEFAULT_DARK_MODE_CLASSES = ["dark", "dark-mode", "theme-dark", "night", "night-mode"];
  const htmlElement = document.documentElement;
  const bodyElement = document.body || document.documentElement;
  for (const className of DEFAULT_DARK_MODE_CLASSES) {
    if (htmlElement.classList.contains(className) || bodyElement?.classList.contains(className)) {
      return true;
    }
  }
  const darkThemeAttribute = htmlElement.getAttribute("data-theme");
  if (darkThemeAttribute?.toLowerCase().includes("dark")) {
    return true;
  }
  return false;
}
__name(hasDarkModeClass, "hasDarkModeClass");
function parseRgbColor(colorString) {
  const rgbMatch = /rgba?\((\d+),\s*(\d+),\s*(\d+)/.exec(colorString);
  if (!rgbMatch) {
    return null;
  }
  return {
    r: parseInt(rgbMatch[1]),
    g: parseInt(rgbMatch[2]),
    b: parseInt(rgbMatch[3])
  };
}
__name(parseRgbColor, "parseRgbColor");
function isColorDark(colorString, threshold = 128) {
  if (!colorString || colorString === "transparent" || colorString.startsWith("rgba(0, 0, 0, 0)")) {
    return false;
  }
  const rgb = parseRgbColor(colorString);
  if (!rgb) {
    return false;
  }
  const luminance = 0.299 * rgb.r + 0.587 * rgb.g + 0.114 * rgb.b;
  return luminance < threshold;
}
__name(isColorDark, "isColorDark");
function isBackgroundDark() {
  const htmlStyle = window.getComputedStyle(document.documentElement);
  const bodyStyle = window.getComputedStyle(document.body || document.documentElement);
  const htmlBgColor = htmlStyle.backgroundColor;
  const bodyBgColor = bodyStyle.backgroundColor;
  if (isColorDark(bodyBgColor)) {
    return true;
  } else if (bodyBgColor === "transparent" || bodyBgColor.startsWith("rgba(0, 0, 0, 0)")) {
    return isColorDark(htmlBgColor);
  }
  return false;
}
__name(isBackgroundDark, "isBackgroundDark");
function isPageDark() {
  try {
    if (hasDarkModeClass()) {
      return true;
    }
    if (isBackgroundDark()) {
      return true;
    }
    return false;
  } catch (error) {
    console.warn("Error determining if page is dark:", error);
    return false;
  }
}
__name(isPageDark, "isPageDark");
const wrapper = "_wrapper_1ooyb_1";
const visible = "_visible_1ooyb_11";
const styles = {
  wrapper,
  visible
};
const cursor = "_cursor_1dgwb_2";
const cursorBorder = "_cursorBorder_1dgwb_10";
const cursorFilling = "_cursorFilling_1dgwb_25";
const cursorRipple = "_cursorRipple_1dgwb_39";
const clicking = "_clicking_1dgwb_57";
const cursorStyles = {
  cursor,
  cursorBorder,
  cursorFilling,
  cursorRipple,
  clicking
};
const _SimulatorMask = class _SimulatorMask {
  constructor() {
    __privateAdd(this, _SimulatorMask_instances);
    __publicField(this, "shown", false);
    __publicField(this, "wrapper", document.createElement("div"));
    __publicField(this, "motion", null);
    __privateAdd(this, _cursor, document.createElement("div"));
    __privateAdd(this, _currentCursorX, 0);
    __privateAdd(this, _currentCursorY, 0);
    __privateAdd(this, _targetCursorX, 0);
    __privateAdd(this, _targetCursorY, 0);
    this.wrapper.id = "page-agent-runtime_simulator-mask";
    this.wrapper.className = styles.wrapper;
    this.wrapper.setAttribute("data-browser-use-ignore", "true");
    this.wrapper.setAttribute("data-page-agent-ignore", "true");
    try {
      const motion = new Motion({
        mode: isPageDark() ? "dark" : "light",
        styles: { position: "absolute", inset: "0" }
      });
      this.motion = motion;
      this.wrapper.appendChild(motion.element);
      motion.autoResize(this.wrapper);
    } catch (e) {
      console.warn("[SimulatorMask] Motion overlay unavailable:", e);
    }
    this.wrapper.addEventListener("click", (e) => {
      e.stopPropagation();
      e.preventDefault();
    });
    this.wrapper.addEventListener("mousedown", (e) => {
      e.stopPropagation();
      e.preventDefault();
    });
    this.wrapper.addEventListener("mouseup", (e) => {
      e.stopPropagation();
      e.preventDefault();
    });
    this.wrapper.addEventListener("mousemove", (e) => {
      e.stopPropagation();
      e.preventDefault();
    });
    this.wrapper.addEventListener("wheel", (e) => {
      e.stopPropagation();
      e.preventDefault();
    });
    this.wrapper.addEventListener("keydown", (e) => {
      e.stopPropagation();
      e.preventDefault();
    });
    this.wrapper.addEventListener("keyup", (e) => {
      e.stopPropagation();
      e.preventDefault();
    });
    __privateMethod(this, _SimulatorMask_instances, createCursor_fn).call(this);
    document.body.appendChild(this.wrapper);
    __privateMethod(this, _SimulatorMask_instances, moveCursorToTarget_fn).call(this);
    window.addEventListener("PageAgent::MovePointerTo", (event) => {
      const { x, y } = event.detail;
      this.setCursorPosition(x, y);
    });
    window.addEventListener("PageAgent::ClickPointer", (event) => {
      this.triggerClickAnimation();
    });
  }
  setCursorPosition(x, y) {
    __privateSet(this, _targetCursorX, x);
    __privateSet(this, _targetCursorY, y);
  }
  triggerClickAnimation() {
    __privateGet(this, _cursor).classList.remove(cursorStyles.clicking);
    void __privateGet(this, _cursor).offsetHeight;
    __privateGet(this, _cursor).classList.add(cursorStyles.clicking);
  }
  show() {
    if (this.shown) return;
    this.shown = true;
    this.motion?.start();
    this.motion?.fadeIn();
    this.wrapper.classList.add(styles.visible);
    __privateSet(this, _currentCursorX, window.innerWidth / 2);
    __privateSet(this, _currentCursorY, window.innerHeight / 2);
    __privateSet(this, _targetCursorX, __privateGet(this, _currentCursorX));
    __privateSet(this, _targetCursorY, __privateGet(this, _currentCursorY));
    __privateGet(this, _cursor).style.left = `${__privateGet(this, _currentCursorX)}px`;
    __privateGet(this, _cursor).style.top = `${__privateGet(this, _currentCursorY)}px`;
  }
  hide() {
    if (!this.shown) return;
    this.shown = false;
    this.motion?.fadeOut();
    this.motion?.pause();
    __privateGet(this, _cursor).classList.remove(cursorStyles.clicking);
    setTimeout(() => {
      this.wrapper.classList.remove(styles.visible);
    }, 800);
  }
  dispose() {
    this.motion?.dispose();
    this.wrapper.remove();
  }
};
_cursor = new WeakMap();
_currentCursorX = new WeakMap();
_currentCursorY = new WeakMap();
_targetCursorX = new WeakMap();
_targetCursorY = new WeakMap();
_SimulatorMask_instances = new WeakSet();
createCursor_fn = /* @__PURE__ */ __name(function() {
  __privateGet(this, _cursor).className = cursorStyles.cursor;
  const rippleContainer = document.createElement("div");
  rippleContainer.className = cursorStyles.cursorRipple;
  __privateGet(this, _cursor).appendChild(rippleContainer);
  const fillingLayer = document.createElement("div");
  fillingLayer.className = cursorStyles.cursorFilling;
  __privateGet(this, _cursor).appendChild(fillingLayer);
  const borderLayer = document.createElement("div");
  borderLayer.className = cursorStyles.cursorBorder;
  __privateGet(this, _cursor).appendChild(borderLayer);
  this.wrapper.appendChild(__privateGet(this, _cursor));
}, "#createCursor");
moveCursorToTarget_fn = /* @__PURE__ */ __name(function() {
  const newX = __privateGet(this, _currentCursorX) + (__privateGet(this, _targetCursorX) - __privateGet(this, _currentCursorX)) * 0.2;
  const newY = __privateGet(this, _currentCursorY) + (__privateGet(this, _targetCursorY) - __privateGet(this, _currentCursorY)) * 0.2;
  const xDistance = Math.abs(newX - __privateGet(this, _targetCursorX));
  if (xDistance > 0) {
    if (xDistance < 2) {
      __privateSet(this, _currentCursorX, __privateGet(this, _targetCursorX));
    } else {
      __privateSet(this, _currentCursorX, newX);
    }
    __privateGet(this, _cursor).style.left = `${__privateGet(this, _currentCursorX)}px`;
  }
  const yDistance = Math.abs(newY - __privateGet(this, _targetCursorY));
  if (yDistance > 0) {
    if (yDistance < 2) {
      __privateSet(this, _currentCursorY, __privateGet(this, _targetCursorY));
    } else {
      __privateSet(this, _currentCursorY, newY);
    }
    __privateGet(this, _cursor).style.top = `${__privateGet(this, _currentCursorY)}px`;
  }
  requestAnimationFrame(() => __privateMethod(this, _SimulatorMask_instances, moveCursorToTarget_fn).call(this));
}, "#moveCursorToTarget");
__name(_SimulatorMask, "SimulatorMask");
let SimulatorMask = _SimulatorMask;
export {
  SimulatorMask
};
