// @page-agent/llms/dist/lib/page-agent-llms.js
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
import * as z from "zod/v4";
import chalk from "chalk";
const InvokeErrorType = {
  // Retryable
  NETWORK_ERROR: "network_error",
  // Network error, retry
  RATE_LIMIT: "rate_limit",
  // Rate limit, retry
  SERVER_ERROR: "server_error",
  // 5xx, retry
  NO_TOOL_CALL: "no_tool_call",
  // Model did not call tool
  INVALID_TOOL_ARGS: "invalid_tool_args",
  // Tool args don't match schema
  TOOL_EXECUTION_ERROR: "tool_execution_error",
  // Tool execution error
  UNKNOWN: "unknown",
  // Non-retryable
  AUTH_ERROR: "auth_error",
  // Authentication failed
  CONTEXT_LENGTH: "context_length",
  // Prompt too long
  CONTENT_FILTER: "content_filter"
  // Content filtered
};
const _InvokeError = class _InvokeError extends Error {
  type;
  retryable;
  statusCode;
  /* raw error (provided if this error is caused by another error) */
  rawError;
  /* raw response from the API (provided if this error is caused by an API calling) */
  rawResponse;
  constructor(type, message, rawError, rawResponse) {
    super(message);
    this.name = "InvokeError";
    this.type = type;
    this.retryable = this.isRetryable(type, rawError);
    this.rawError = rawError;
    this.rawResponse = rawResponse;
  }
  isRetryable(type, rawError) {
    const isAbortError = rawError?.name === "AbortError";
    if (isAbortError) return false;
    const retryableTypes = [
      InvokeErrorType.NETWORK_ERROR,
      InvokeErrorType.RATE_LIMIT,
      InvokeErrorType.SERVER_ERROR,
      InvokeErrorType.NO_TOOL_CALL,
      InvokeErrorType.INVALID_TOOL_ARGS,
      InvokeErrorType.TOOL_EXECUTION_ERROR,
      InvokeErrorType.UNKNOWN
    ];
    return retryableTypes.includes(type);
  }
};
__name(_InvokeError, "InvokeError");
let InvokeError = _InvokeError;
const debug = console.debug.bind(console, chalk.gray("[LLM]"));
function zodToOpenAITool(name, tool) {
  return {
    type: "function",
    function: {
      name,
      description: tool.description,
      parameters: z.toJSONSchema(tool.inputSchema, { target: "openapi-3.0" })
    }
  };
}
__name(zodToOpenAITool, "zodToOpenAITool");
function modelPatch(body) {
  const model = body.model || "";
  if (!model) return body;
  const modelName = normalizeModelName(model);
  if (modelName.startsWith("qwen")) {
    debug("Applying Qwen patch: use higher temperature for auto fixing");
    body.temperature = Math.max(body.temperature || 0, 1);
    body.enable_thinking = false;
  }
  if (modelName.startsWith("claude")) {
    debug("Applying Claude patch: disable thinking");
    body.thinking = { type: "disabled" };
    if (body.tool_choice === "required") {
      debug('Applying Claude patch: convert tool_choice "required" to { type: "any" }');
      body.tool_choice = { type: "any" };
    } else if (body.tool_choice?.function?.name) {
      debug("Applying Claude patch: convert tool_choice format");
      body.tool_choice = { type: "tool", name: body.tool_choice.function.name };
    }
  }
  if (modelName.startsWith("grok")) {
    debug("Applying Grok patch: removing tool_choice");
    delete body.tool_choice;
    debug("Applying Grok patch: disable reasoning and thinking");
    body.thinking = { type: "disabled", effort: "minimal" };
    body.reasoning = { enabled: false, effort: "low" };
  }
  if (modelName.startsWith("gpt")) {
    debug("Applying GPT patch: set verbosity to low");
    body.verbosity = "low";
    if (modelName.startsWith("gpt-52")) {
      debug("Applying GPT-52 patch: disable reasoning");
      body.reasoning_effort = "none";
    } else if (modelName.startsWith("gpt-51")) {
      debug("Applying GPT-51 patch: disable reasoning");
      body.reasoning_effort = "none";
    } else if (modelName.startsWith("gpt-54")) {
      debug(
        "Applying GPT-5.4 patch: skip reasoning_effort because chat/completions rejects it with function tools"
      );
      delete body.reasoning_effort;
    } else if (modelName.startsWith("gpt-5-mini")) {
      debug("Applying GPT-5-mini patch: set reasoning effort to low, temperature to 1");
      body.reasoning_effort = "low";
      body.temperature = 1;
    } else if (modelName.startsWith("gpt-5")) {
      debug("Applying GPT-5 patch: set reasoning effort to low");
      body.reasoning_effort = "low";
    }
  }
  if (modelName.startsWith("gemini")) {
    debug("Applying Gemini patch: set reasoning effort to minimal");
    body.reasoning_effort = "minimal";
  }
  if (modelName.startsWith("minimax")) {
    debug("Applying MiniMax patch: clamp temperature to (0, 1]");
    body.temperature = Math.max(body.temperature || 0, 0.01);
    if (body.temperature > 1) body.temperature = 1;
    delete body.parallel_tool_calls;
  }
  return body;
}
__name(modelPatch, "modelPatch");
function normalizeModelName(modelName) {
  let normalizedName = modelName.toLowerCase();
  if (normalizedName.includes("/")) {
    normalizedName = normalizedName.split("/")[1];
  }
  normalizedName = normalizedName.replace(/_/g, "");
  normalizedName = normalizedName.replace(/\./g, "");
  return normalizedName;
}
__name(normalizeModelName, "normalizeModelName");
const _OpenAIClient = class _OpenAIClient {
  config;
  fetch;
  constructor(config) {
    this.config = config;
    this.fetch = config.customFetch;
  }
  async invoke(messages, tools, abortSignal, options) {
    const openaiTools = Object.entries(tools).map(([name, t]) => zodToOpenAITool(name, t));
    let toolChoice = "required";
    if (options?.toolChoiceName && !this.config.disableNamedToolChoice) {
      toolChoice = { type: "function", function: { name: options.toolChoiceName } };
    }
    const requestBody = {
      model: this.config.model,
      temperature: this.config.temperature,
      messages,
      tools: openaiTools,
      parallel_tool_calls: false,
      tool_choice: toolChoice
    };
    modelPatch(requestBody);
    let response;
    try {
      response = await this.fetch(`${this.config.baseURL}/chat/completions`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          ...this.config.apiKey && { Authorization: `Bearer ${this.config.apiKey}` }
        },
        body: JSON.stringify(requestBody),
        signal: abortSignal
      });
    } catch (error) {
      const isAbortError = error?.name === "AbortError";
      const errorMessage = isAbortError ? "Network request aborted" : "Network request failed";
      if (!isAbortError) console.error(error);
      throw new InvokeError(InvokeErrorType.NETWORK_ERROR, errorMessage, error);
    }
    if (!response.ok) {
      const errorData = await response.json().catch();
      const errorMessage = errorData.error?.message || response.statusText;
      if (response.status === 401 || response.status === 403) {
        throw new InvokeError(
          InvokeErrorType.AUTH_ERROR,
          `Authentication failed: ${errorMessage}`,
          errorData
        );
      }
      if (response.status === 429) {
        throw new InvokeError(
          InvokeErrorType.RATE_LIMIT,
          `Rate limit exceeded: ${errorMessage}`,
          errorData
        );
      }
      if (response.status >= 500) {
        throw new InvokeError(
          InvokeErrorType.SERVER_ERROR,
          `Server error: ${errorMessage}`,
          errorData
        );
      }
      throw new InvokeError(
        InvokeErrorType.UNKNOWN,
        `HTTP ${response.status}: ${errorMessage}`,
        errorData
      );
    }
    const data = await response.json();
    const choice = data.choices?.[0];
    if (!choice) {
      throw new InvokeError(InvokeErrorType.UNKNOWN, "No choices in response", data);
    }
    switch (choice.finish_reason) {
      case "tool_calls":
      case "function_call":
      // gemini
      case "stop":
        break;
      case "length":
        throw new InvokeError(
          InvokeErrorType.CONTEXT_LENGTH,
          "Response truncated: max tokens reached",
          void 0,
          data
        );
      case "content_filter":
        throw new InvokeError(
          InvokeErrorType.CONTENT_FILTER,
          "Content filtered by safety system",
          void 0,
          data
        );
      default:
        throw new InvokeError(
          InvokeErrorType.UNKNOWN,
          `Unexpected finish_reason: ${choice.finish_reason}`,
          void 0,
          data
        );
    }
    const normalizedData = options?.normalizeResponse ? options.normalizeResponse(data) : data;
    const normalizedChoice = normalizedData.choices?.[0];
    const toolCallName = normalizedChoice?.message?.tool_calls?.[0]?.function?.name;
    if (!toolCallName) {
      throw new InvokeError(
        InvokeErrorType.NO_TOOL_CALL,
        "No tool call found in response",
        void 0,
        data
      );
    }
    const tool = tools[toolCallName];
    if (!tool) {
      throw new InvokeError(
        InvokeErrorType.UNKNOWN,
        `Tool "${toolCallName}" not found in tools`,
        void 0,
        data
      );
    }
    const argString = normalizedChoice.message?.tool_calls?.[0]?.function?.arguments;
    if (!argString) {
      throw new InvokeError(
        InvokeErrorType.INVALID_TOOL_ARGS,
        "No tool call arguments found",
        void 0,
        data
      );
    }
    let parsedArgs;
    try {
      parsedArgs = JSON.parse(argString);
    } catch (error) {
      throw new InvokeError(
        InvokeErrorType.INVALID_TOOL_ARGS,
        "Failed to parse tool arguments as JSON",
        error,
        data
      );
    }
    const validation = tool.inputSchema.safeParse(parsedArgs);
    if (!validation.success) {
      console.error(z.prettifyError(validation.error));
      throw new InvokeError(
        InvokeErrorType.INVALID_TOOL_ARGS,
        "Tool arguments validation failed",
        validation.error,
        data
      );
    }
    const toolInput = validation.data;
    let toolResult;
    try {
      toolResult = await tool.execute(toolInput);
    } catch (e) {
      throw new InvokeError(
        InvokeErrorType.TOOL_EXECUTION_ERROR,
        `Tool execution failed: ${e.message}`,
        e,
        data
      );
    }
    return {
      toolCall: {
        name: toolCallName,
        args: toolInput
      },
      toolResult,
      usage: {
        promptTokens: data.usage?.prompt_tokens ?? 0,
        completionTokens: data.usage?.completion_tokens ?? 0,
        totalTokens: data.usage?.total_tokens ?? 0,
        cachedTokens: data.usage?.prompt_tokens_details?.cached_tokens,
        reasoningTokens: data.usage?.completion_tokens_details?.reasoning_tokens
      },
      rawResponse: data,
      rawRequest: requestBody
    };
  }
};
__name(_OpenAIClient, "OpenAIClient");
let OpenAIClient = _OpenAIClient;
const LLM_MAX_RETRIES = 2;
const DEFAULT_TEMPERATURE = 0.7;
function parseLLMConfig(config) {
  if (!config.baseURL || !config.model) {
    throw new Error(
      "[PageAgent] LLM configuration required. Please provide: baseURL, model. See: https://alibaba.github.io/page-agent/docs/features/models"
    );
  }
  return {
    baseURL: config.baseURL,
    model: config.model,
    apiKey: config.apiKey || "",
    temperature: config.temperature ?? DEFAULT_TEMPERATURE,
    maxRetries: config.maxRetries ?? LLM_MAX_RETRIES,
    disableNamedToolChoice: config.disableNamedToolChoice ?? false,
    customFetch: (config.customFetch ?? fetch).bind(globalThis)
    // fetch will be illegal unless bound
  };
}
__name(parseLLMConfig, "parseLLMConfig");
const _LLM = class _LLM extends EventTarget {
  config;
  client;
  constructor(config) {
    super();
    this.config = parseLLMConfig(config);
    this.client = new OpenAIClient(this.config);
  }
  /**
   * - call llm api *once*
   * - invoke tool call *once*
   * - return the result of the tool
   */
  async invoke(messages, tools, abortSignal, options) {
    return await withRetry(
      async () => {
        if (abortSignal.aborted) throw new Error("AbortError");
        const result = await this.client.invoke(messages, tools, abortSignal, options);
        return result;
      },
      // retry settings
      {
        maxRetries: this.config.maxRetries,
        onRetry: /* @__PURE__ */ __name((attempt) => {
          this.dispatchEvent(
            new CustomEvent("retry", { detail: { attempt, maxAttempts: this.config.maxRetries } })
          );
        }, "onRetry"),
        onError: /* @__PURE__ */ __name((error) => {
          this.dispatchEvent(new CustomEvent("error", { detail: { error } }));
        }, "onError")
      }
    );
  }
};
__name(_LLM, "LLM");
let LLM = _LLM;
async function withRetry(fn, settings) {
  let attempt = 0;
  let lastError = null;
  while (attempt <= settings.maxRetries) {
    if (attempt > 0) {
      settings.onRetry(attempt);
      await new Promise((resolve) => setTimeout(resolve, 100));
    }
    try {
      return await fn();
    } catch (error) {
      if (error?.rawError?.name === "AbortError") throw error;
      console.error(error);
      settings.onError(error);
      if (error instanceof InvokeError && !error.retryable) throw error;
      lastError = error;
      attempt++;
      await new Promise((resolve) => setTimeout(resolve, 100));
    }
  }
  throw lastError;
}
__name(withRetry, "withRetry");
export {
  InvokeError,
  InvokeErrorType,
  LLM,
  parseLLMConfig
};
